content/facets-endpoint/index.md

```diff
@@ -0,0 +1,93 @@
+---
+rfc: 0022
+start_date: 2018-09-28
+decision_date:
+pr: openregisters/registers-rfc#39
+status: draft
+---
+
+# Facets endpoint
+
+## Summary
+
+This RFC proposes to change the endpoint for the records' “facets” to simplify
+path handling.
+
+## Motivation
+
+Currently the REST API for facets uses a path structure that overlaps with the
+standard record structure. This makes path handling (e.g. analytics, logging)
+harder than it needs to be. Observation suggests that users can get confused
+of what is the resource about.
+
+
+## Explanation
+
+For context, the current endpoint for facets is:
+
+```
+GET /records/{attribute-name}/{attribute-value}
+```
+
+The record endpoint is:
+
+```
+GET /records/{key}
+```
+
+And the records endpoint is:
+
+```
+GET /records
+```
+
+
+This RFC proposes to change the facets endpoint to extend the records endpoint
+with a query string to make more obvious that a “facet” is filter on the
+records collection.
+
+```
+GET /records{?name, value}
+```
+
+Where `{?name}` is required if `{?value}` is informed.
+
+
+### Filter by strict value match
+
+```
+GET /records?name={attribute-name}&value={attribute-value}
```

Comment:
A limitation of this is we can only filter using one attribute at a time, are we ok with that?

Also many registers have a "name" field so I don't know if it would be confusing to use `name` as a param here.

Another alternative would be something like

```
/records?filter_{attribute-name}={attribute-value}
```

Although you'd need to use a special value to implement the next use case.

IMO GOV.UK's search API implemented this quite well - their [filtering API](https://docs.publishing.service.gov.uk/apis/search/faceted-search.html#filtering-parameters) can either include or exclude stuff, and it can also express null values and date ranges.

Reply:
> A limitation of this is we can only filter using one attribute at a time, are we ok with that? I'm unsure at this point what we want to do with facets, and whether we would use them to address any of the use cases indexes were designed to address.

The proposal changes the way we interact with facets but it doesn't change the capabilities. Accepting multiple attributes is a deep change that require much more than a change in the API.

> Also many registers have a "name" field so I don't know if it would be confusing to use name as a param here.

I don't agree with being confusing (or more or less confusing than any other combination like the one you suggest). But it's just my opinion. I don't  have proper backup to sustain it beyond that, do you?

> IMO GOV.UK's search API implemented this quite well - their filtering API can either include or exclude stuff, and it can also express null values and date ranges.

I'm going to disagree here. GOV.UK has a different need, and it looks like they went for something on the lines of the Elasticsearch DSL which is overkill for us.